src/Compilation.zig

```diff
@@ -1861,16 +1861,69 @@ pub fn create(gpa: Allocator, arena: Allocator, options: CreateOptions) !*Compil
                 try comp.queueJob(.libtsan);
             }
 
-            if (target.isMinGW() and comp.config.any_non_single_threaded) {
-                // LLD might drop some symbols as unused during LTO and GCing, therefore,
-                // we force mark them for resolution here.
+            // https://github.com/llvm/llvm-project/issues/43698#issuecomment-2542660611
+            //
+            // This only seems to affect Windows/COFF, for whatever reason.
+            if (target.os.tag == .windows) {
+                if (comp.config.any_non_single_threaded) {
+                    try comp.force_undefined_symbols.put(comp.gpa, if (target.cpu.arch == .x86) "__tls_index" else "_tls_index", {});
```

Comment:
Please don't add things to this data structure. It looks like a hack and I don't understand why it is needed. The doc comments on the field are insufficient to justify its existence.

Reply:
It's a workaround for an LLVM/LLD deficiency.

Per https://github.com/llvm/llvm-project/issues/43698, LLD has an incomplete view of the external symbols that LLVM backends might emit references to because LLVM's `RuntimeLibcalls.def` is incomplete (containing mostly target-agnostic symbols), and there is not yet consensus on how to fix that. It used to be that the LTO infrastructure would query the `TargetLowering` interface for the *actual* libcall names in effect at lowering time, but in the new LTO infrastructure, that is no longer possible because the querying needs to happen in LLD's driver - long before a `TargetLowering` instance can be obtained.

Here's the relevant LLD code:

* https://github.com/llvm/llvm-project/blob/99354f968f64659cbad5c82b0301d851ae54f057/lld/COFF/Driver.cpp#L2606-L2616
* https://github.com/llvm/llvm-project/blob/99354f968f64659cbad5c82b0301d851ae54f057/lld/COFF/SymbolTable.cpp#L914-L927
* https://github.com/llvm/llvm-project/blob/99354f968f64659cbad5c82b0301d851ae54f057/lld/COFF/Driver.cpp#L416-L418

As you can see here, if those symbols *were* properly declared as libcalls, LLD would actually do for them the equivalent of what we do here by passing `-include:<symbol>` through the `force_undefined_symbols` mechanism.

I haven't dug deep enough into the ELF portion of LLD to be able to say why it seems completely unaffected by this problem. LLD's symbol resolution strategy for COFF seems fairly different from other formats, though.

I think that the main reason we run into this problem but e.g. Clang doesn't is that it's very rare to LTO `compiler_rt.lib` (`_chkstk`, etc) and `mingw32.lib` (`_tls_index`, etc). I'm fairly sure a normal Clang toolchain doesn't do that.

Reply:
@andrewrk do you find the above explanation sufficient? If so, I can try to reproduce it in a code comment as well.

I can also just leave this out if you still think it's too hacky, although one way or another, we have to work around that deficiency. One alternative that is simpler is to disable LTO altogether for Zig-provided libraries like compiler-rt and MinGW when targeting Windows. I didn't propose that here, though, because that's painting with a much broader brush.

Reply:
The original motivation for introducing `force_undefined_symbols` which under-the-hood resolves to `-include` for COFF linker, `-u` for ELF/MachO. Next, specifically for COFF, it was used as a workaround hack around improper dropping of `__tls_index` symbol during LTO/GC passes. Having said that I never actually got round to figuring out if this is indeed a problem with the linker marking the symbol dead *or* mingw not setting the symbol flags correctly (I seem to remember it being defined by mingw, and do remember tweaking its definition which seemed to have fixed the problem also). Anyhow, doing the same for compiler-rt symbols seems sus especially since we control its generation. Maybe it would be better to repro clang's behaviour here and not subject compiler-rt to LTO? Or did I misunderstand your suggestion?

Reply:
> Having said that I never actually got round to figuring out if this is indeed a problem with the linker marking the symbol dead _or_ mingw not setting the symbol flags correctly (I seem to remember it being defined by mingw, and do remember tweaking its definition which seemed to have fixed the problem also).

I don't think it has anything to do with MinGW really. Consider what we know for sure:

* It only happens with LLD's COFF linker.
* It only happens with link-time GC and LTO.
* It only happens for symbols that:
    * are defined in bitcode files which are themselves contained within static library archives (so `compiler_rt.lib` and `mingw32.lib` in our case).
    * are not properly declared in LLVM's `RuntimeLibcalls.def` (or some moral equivalent).
    * are referenced only by machine code sequences created in LLVM's backends.
    * are never referenced explicitly by LLVM IR.

When you consider all of these factors, it makes sense that, for Zig, it affects `_tls_index` (defined in `mingw32.lib`) and `_chkstk`, etc (defined in `compiler_rt.lib`) and occurs both for MinGW and MSVC targets. We LTO both of these libraries when LTO is enabled. And it makes sense that typical Clang toolchains don't have this problem because they build compiler-rt and MinGW as normal static libraries rather than bitcode, and only LTO 'user-level' code.

> Maybe it would be better to repro clang's behaviour here and not subject compiler-rt to LTO? Or did I misunderstand your suggestion?

Right, that is one possible solution to this. It's either we mark these symbols as undefined to work around the LLVM/LLD deficiency, or we don't LTO Zig's compiler-provided libraries when targeting COFF. My preference is very slightly in favor of marking the symbols undefined so we don't completely lose LTO for the entirety of these libraries, but I wouldn't lose sleep over it if we decided to just disable LTO for them.

Reply:
Makes sense, thanks for an explanation!